src/main/java/com/meta/cp4m/configuration/RootConfiguration.java

```diff
@@ -137,7 +144,22 @@ private <T extends Message> Service<T> createService(
     } else {
       store = new NullStore<T>();
     }
-    return new Service<>(store, handler, plugin, serviceConfig.webhookPath());
+
+    PreProcessor<T> preProcessor;
+    List<PreProcessor<T>> preProcessorsList = new ArrayList<>();
+
+    if(serviceConfig.preProcessors() != null){
+      String[] preProcessorNames = serviceConfig.preProcessors();
+      for (String i : preProcessorNames) {
+        preProcessor = preProcessors.get(i).toPreProcessor();
+        preProcessorsList.add(preProcessor);
+      }
+
+      return new Service<>(store, handler, plugin, preProcessorsList, serviceConfig.webhookPath());
+    } else {
+      // TODO: flow never goes into else. Need to debug why method inherits container annotation
```

Comment:
I have a NonNull annotation on everything in the cp4m package by default (it can be explicitly overridden) https://github.com/facebookincubator/CP4M/blob/main/src/main/java/com/meta/cp4m/package-info.java#L9

However annotations don't nullability annotation do not effect runtime values they're just there to give IDE and compile time warnings. I can work with you to understand how you're not getting null here, however I added a suggestion above to avoid the null ever occurring (see Item 54 of "Effective Java",  "Return empty collections or arrays, not nulls")

Reply:
I don't see this suggested link, any chance it's not attached?